src/tests/tables.test.js

```diff
@@ -3,20 +3,94 @@ const glob = require("glob");
+        fileIndices[filename] = new Set();
+      }
+
+      if (fileIndices[filename].has(entry.index)) {
+        errors.push(`${filename}: Index '${entry.index}' already exists.`);
+      }
+
+      fileIndices[filename].add(entry.index);
+    });
+
+    expect(errors).toEqual([]);
+  });
+});
+
+describe("api references", () => {
+  it("should not contain broken links", () => {
     let errors = [];
     let files = glob.sync("src/*.json");
 
-    for (const file of files) {
-      const fileText = fs.readFileSync(file, "utf8");
-      const fileJSON = JSON.parse(fileText);
-      const indices = new Set();
-      fileJSON.forEach((entry) => {
-        if (indices.has(entry.index)) {
-          errors.push(`${file}: Index '${entry.index}' already exists.`);
+    let resources = {};
+
+    forEachFile((filename, entry) => {
+      if (entry.url !== undefined) {
+        if (entry.index === undefined) {
+          errors.push(`${filename}: Entry with URL '${entry.url}' should have an index.`);
         }
-        indices.add(entry.index);
-      });
-    }
+
+        resources[entry.url] = { index: entry.index, name: entry.name };
+      }
+    });
+
+    forEachFileRecursive((filename, entry) => {
+      if (entry.hasOwnProperty("url")) {
+        if (resources[entry.url] === undefined) {
+          errors.push(`${filename}: URL '${entry.url}' not found.`);
+        } else {
+          if (resources[entry.url].name !== undefined && resources[entry.url].name !== entry.name) {
+            errors.push(`${filename}: Name mismatch for reference to '${entry.url}', '${entry.name}' should be '${resources[entry.url].name}'`);
+          }
+
+          if (entry.index !== undefined && resources[entry.url].index !== entry.index) {
+            errors.push(`${filename}: Index mismatch for reference to '${entry.url}', '${entry.index}' should be '${resources[entry.url].index}'`);
+          }
+        }
+      }
+    });
+
     expect(errors).toEqual([]);
   });
 });
+
+/**
+ * Calls the callback for top-level objects/arrays in all JSON files.
+ *
+ * @param (function(string, object)) callback Called with filename and each
+ *     top-level entry.
+ */
+const forEachFile = (callback) => {
+    let filenames = glob.sync("src/*.json");
+
+    for (const filename of filenames) {
+      const fileText = fs.readFileSync(filename, "utf8");
+      const fileJSON = JSON.parse(fileText);
+      fileJSON.forEach((entry) => callback(filename, entry));
+    }
+};
+
+/**
+ * Calls the callback for all non-top-level objects/arrays in all JSON files.
+ * Does not overlap with entries from forEachFile().
+ *
+ * @param (function(string, object)) callback Called with filename and each
+ *     sub-top-level entry.
+ */
+const forEachFileRecursive = (callback) => {
+  const recurse = (filename, entry) => {
+    for (const property in entry) {
+      if (typeof entry[property] === "object" && entry[property] !== null) {
+        callback(filename, entry[property]);
```

Comment:
Should this callback be outside of the `for` loop? As it stands, it seems like it will skip the top level field? While that is what you're after, I think it limits the reuse of this method?

Reply:
Maybe. My gut feeling is that having `forEachFile()` and `forEachFileRecursive()` as non-overlapping complements will be valuable, rather than having to check if you're dealing with a top-level entry or a reference.

Reply:
Mainly, it feels confusing if the top level is skipped for the recursive? What I would expect `forEachFileRecursive` to run the callback on every level and not every non-top level?

Reply:
Fair. Rather than implementing that behaviour and then immediately working around it in the only place where the code is used, I've split out the recursive code as suggested and recombined it with `forEachFile()` in the new test. `recurseInObject()` still has the same behaviour of not calling the callback with the top-level object, but I feel like that's more intuitive now that the object is coming directly from the calling code. I'm not married to that decision, though.

Maybe more problematic, it still only calls the callback for each *object* (or array), not other property types, which suits my use case just fine but might be less intuitive. Still, having your callback invoked with only `null` or `false` isn't very helpful without path information and/or the ability to mutate data in-place, so I'm inclined to let that use case lie and deal with it as needed.